### ISSUE TITLE
Conformer training from scratch not achieving the reported WER

### ISSUE THREAD

**Issue by kssteven418** [NVIDIA/NeMo#3288]
Hello! Thanks for providing such a nice toolkit.

I am currently training conformer-small on librispeech-960hr from scratch, following the training recipe and instructions in `examples/asr/conf/conformer/conformer_ctc_bpe.yaml`.
However, I'm not able to reproduce the WER reported in this [webpage](https://catalog.ngc.nvidia.com/orgs/nvidia/teams/nemo/models/stt_en_conformer_ctc_small_ls): reported ones are 3.4/8.8, but our WER is only around 4.5/11 on the clean/other datasets with ~300 epochs of training and doesn't seem to decrease anymore.

Here is the list of hyperparameters I'm using:
* d_model, n_heads, n_layers: 176, 16, 16 
* tokenizer: WPE with vocab_size 128 (generated using the script in NeMo)
* spec_augment: freq_masks 2, time_masks 5, freq_width:27, time_width: 0.05
* lr: 5.0, betas: [0.9, 0.98], weight_decay: 1e-3 (adamw)
* warmup_steps: 10000 (noam annealing)
* Batch size: 2048 (using 8 machines, local batch size 16, and accumulate_grad_batches 16, multiplying up to 2048)
All the other hyperparameters are the same as the default.

Here is one of the learning curves (validation WER on dev-clean) from our runs:
<img width="1665" alt="Screen Shot 2021-12-05 at 1 18 53 PM" src="https://user-images.githubusercontent.com/50283958/144764357-9cc2609c-cf8e-4158-8693-5e6a7b7c5bd8.png">

I'm wondering if we should change the hyperparameters to achieve the reported WER values, or if the 300-epoch of training is not enough for the model convergence, or if there's something that we are missing.
Any clues regarding the hyperparameters, the number of training epochs, etc. would be really appreciated!

Thank you.

**Comment by amirgholami** [comment]
Is it possible to share the exact settings that the NEMO team used to train conformer-ctc?

**Comment by BuaaAlban** [comment]
Hi， I am meeting the same problem with you, [issue3211](https://github.com/NVIDIA/NeMo/issues/3211) 
My best WER=0.041 @epoch346, and it seems to be converging very slowly   with batch size 32*4=128

**Comment by amirgholami** [comment]
> Hi， I am meeting the same problem with you, [issue3211](https://github.com/NVIDIA/NeMo/issues/3211)
> My best WER=0.041 @epoch346, and it seems to be converging very slowly with batch size 32*4=128 (2 gpus with one node)
> [conformer_ctc_bpe_small.txt](https://github.com/NVIDIA/NeMo/files/7658888/conformer_ctc_bpe_small.txt)

Yes we also saw your issue. We have used the exact same batch size suggested in the answers but the WER is higher than the checkpoint provided.

Hi @VahidooX Vahid,

It would be great if the Nvidia team could provide the detailed training configuration that they have used.

Thanks so much!
-Amir

**Comment by VahidooX** [comment]
That examples/asr/conf/conformer/conformer_ctc_bpe.yaml is the config we used to train the Conformer-CTC-Large model. You just need to update the configs mentioned in the table you see in the file. The number of heads should be 4 not 16. Unfortunataley Conformer-CTC models for LS need a lot of time to converge to SOTA numbers. Among them the small version is the worst. We trained Conformer-CTC models for 1000 epochs on LS with SP of +-10%. It is equivalent to 3000 epochs on regular LS. However the effectiveness of SP is very limited and in most cases insignificant. If you have just trained for 300 epochs, those ac curacies looks OK, you need significantly more epochs to train. 

Conformer-CTC vs Conformer-Transducer:
Larger models are faster in convergence, maybe it is more time-effective if you train larger models to get better accuracy in fewer epochs, or use Conformer-Transducer model. Transducer models may need more GPU memory to train but they are significantly faster to converge than CTC models. For example a Conformer-Transducer-Large converges to less than 7% on test-other LS in less than 40 epochs. A conformer-CTC large converges to under 8% in around 100 epochs. (All epoch numbers we report on training Conformer on LS are with SP)

Tokenizer: If I remember correctly we used senetencepiece unigram tokenizer with that model but WPE should give the same results.

Some notes on training that model with limited number of GPUs:
We used effective batch size of 2K as we wanted to be able to do so many experiments and parameters fine-tuning. You don't need to use such a large batch size. Actually you can get to better accuracy in fewer epochs with smaller batch sizes as you have more updates to the model. I suggest to use effective batch size of 256 (or 512 at most) which means just 2 grad accumulation in your case. That many grad accumulations you used is not efficient in terms of convergence speed. In this case, you may use lr=2 with weight decay of 0.0 to have faster convergence.


Reproducbility:
I trained them more than 9 months ago, need some time to dig them up. I would check the details to make sure it matches the config file we provided. In the meantime, you may load the pretrained checkpoint from NGC and check it's config in model.cfg. I am also going to train this model with batch size of 256 and 2K again to make sure nothing is broken from the time I trained this model. It may take a week or so to get the results. Would update you guys here with results.

**Comment by kssteven418** [comment]
Hi @VahidooX, I recently found a bug in spec augmentation where time maskings were not performed correctly. (issue: #3298, PR: #3299) I guess this might be one of the reasons we are not achieving the desired WERs. This bug was introduced in May~June when the numba kernels for spec augmentation were first implemented, and I assume that all the publically available checkpoints (and all your runs as well) must have been trained before then.

**Comment by titu1994** [comment]
That bug will not affect any of the trained models available via from_pretrained. The support for the Numba kernel is disabled inside of the containers we use to train the models, so it falls back to the python based method.

**Comment by BuaaAlban** [comment]
Hi @VahidooX , is there any update for your training?
I have trained the model for more than 1000 epochs with effective batch size 256, and it has no improvement in recent 500 epochs. Current best  WER on dev-clean is WER = 3.87

**Comment by VahidooX** [comment]
Sorry for the late update. We had a bug to fix which prevented the experiments to be done. I did some experiments with different tokenizers, different batch sizes, different lr and weight decay to make sure that results are reproducible. Experiments are not done yet but here are some of the initial results. Note that you need to multiply the number of epochs by 3 if you do not use the speed perturbation to match our number of epochs)

+ Senetencepiece unigram is better than WPE (range of 0.3% after 300 epochs)
+ lr=5, wd=1e-3 is better than lr=2, wd=0 for small conformer-ctc (better in range of 0.2% after 300 epochs)
+ WER with SP is slightly better on test-other (less than 0.2%/0.1% for test-other/test-clean after 300 epochs)
+ Tried the model with bs=256 and bs=512 and they converge faster epoch-wise.
+ Model with bs=2K and SP gets to 8.8%/3.4% on test-other/test-clean in 620 epochs (eq to 620*3=1860 without SP)
+ Model with bs=2K and SP gets to 10.6%/4.2% on test-other/test-clean in 100 epochs (eq to 100*3=300 without SP)
+ Model with bs=256 and without SP gets to 10.2%/4.0% on test-other/test-clean in 460 epochs (eq to 100/3~153 with SP)
+ Model with bs=512 and without SP gets to 9.5%/3.7% on test-other/test-clean in 1000 epochs (eq to 1000/3~333 with SP)

Your numbers looks OK to me. You just need more than 1800 epochs without SP to get to 8.8% on test-other LS. I suggest you to switch to Conformer-transducer or a larger Conformer-CTC (maybe medium size). Even medium size (30M) is significantly faster than the small model. You may try fp16 with medium size model to improve memory efficiency. You may get to under 10% test-other in 500 epochs withput SP. Also update your code with the main branch of nemo as you may be using the buggy spec aug. That bug limited the augmentation to just the beginning of an audio.
If you decide to start a new training, I suggest to switch to senetencepiece unigram with vocab size of 128, use batch size of 256 or 512 considering your limited number of GPUs. You may start a new training by initialization of the new model with one of the checkpoints you already trained to make the convergence faster.

**Comment by BuaaAlban** [comment]
Thanks a lot for your kindly reply and detailed exps!  It gave us a lot insights for the training of conformer-CTC small.

I still have some questions about the batch_size and Learning rate.

1. Does the bs we talk about equals to batch_size*num_nodes*gpus*accumulate_grad_batches ? 
           I have tried one Model with bs=32*1*4*16=2048  without SP, the following is the results
                0.0784@epoch49   0.0465@epoch147   0.0453@epoch195  0.0432@269
           I stopped it because it seems converge too slow, will it get to 8.8%/3.4% on test-other/test-clean in 620 epochs like your model?(eq to 620*3=1860 without SP)
           
2. Do we set  lr=5, wd=1e-3  for all batch size (256/512/2k)? 
        Is there any training tips or we just set the params as in the config  and just wait for it to converge?
        Because my model with batch size 256 achieves WER=0.0387 insce epoch 559/718/732, and it didn't get any improvement since then, now the training has reached 1337 epoch with WER=0.0411 on dev-clean dataset (I use dev-clean for val and keep the best 3 on validation set). I doubt whether it will converges to about 3.4% thougn I am still training it and waiting for it to reach 3000 epochs.
        BTW, my training stopped due to some reason and I resumed the training by the parameter  “resume_from_checkpoint: last.ckpt” to continue the training, will this have any effect on the final convergence (like the optimizer or something like this). 
      
3.    Why do you suggest us to try the model with bs=256 and bs=512 and they converge faster epoch-wise ?
      Model with bs=2K and SP gets to 8.8%/3.4% on test-other/test-clean in 620 epochs (eq to 620*3=1860 without SP)
      Model with bs=2K and SP gets to 10.6%/4.2% on test-other/test-clean in 100 epochs (eq to 100*3=300 without SP)
      Model with bs=256 and without SP gets to 10.2%/4.0% on test-other/test-clean in 460 epochs (eq to 100/3~153 with SP)
      Model with bs=256 and without SP gets to 9.5%/3.7% on test-other/test-clean in 1000 epochs (eq to 1000/3~333 with SP)
      
      Does this mean that the Model with bs=512/256 and without SP  will get to  8.8%/3.4% in less than 1860 epochs ? 
      
      As described in question 2, my model with bs=256 without SP seems get worse result than your Model with bs=256 and without SP (at 460 epochs they are similar, but worse at 1000 epoch), could you attach your config file for the 4 experiments if possible so that I could compare mine with yours?
      
     
Thanks again for your kindly guide and the details of training results.

**Comment by VahidooX** [comment]
>>> Does the bs we talk about equals to batch_sizenum_nodesgpusaccumulate_grad_batches ?
Yes, exactly.

>>>I have tried one Model with bs=321416=2048 without SP, the following is the results 0.0784@epoch49 0.0465@epoch147 0.0453@epoch195 0.0432@269. 
In my similar experiment, I get similar WERs of 4.6@195 and 4.3@269. Mine was converging slightly faster than yours at the beginning but yours looks like to catch up. It can be due to the difference between our max_epoch as it affects the learning curve with Noam scheduler (set it to 3000 for without SP). 

>>>I stopped it because it seems converge too slow, will it get to 8.8%/3.4% on test-other/test-clean in 620 epochs like your model?(eq to 620*3=1860 without SP)
I compared the run with SP and the one without SP again. It looks like while they are similar at the beginning but after a certain epoch the run with SP converges faster to better results. For example the one with SP converges to 8.8% after 650 epochs (~1950 epoch without SP) but the run without SP converges to 9.1% after 1800 epochs. So if you train for more than 2000 epochs, you would be likely to get under 9%. The run without SP is not finished yet, when it is done I can share the exact WER it converged. The run with SP converged to 8.7% after 1000 epochs.

>>>Do we set lr=5, wd=1e-3 for all batch size (256/512/2k)?
I used the same config for all batch sizes. You may get better results with other configs for smaller batch sizes but I think it would not be dramatic based on my experience.

>>>Because my model with batch size 256 achieves WER=0.0387 at epoch 559/718/732, and it didn't get any improvement since then, now the training has reached 1337 epoch with WER=0.0411 on dev-clean dataset (I use dev-clean for val and keep the best 3 on validation set).
My run with bs=256 without SP reaches 3.8% @470 and it is still at 3.8% at epoch 741. On clean set, it would even be slower to converge. So you just need to train more. Do you mean that the WER is getting worse and you have an overfitting issue?

>>>BTW, my training stopped due to some reason and I resumed the training by the parameter “resume_from_checkpoint: last.ckpt” to continue the training, will this have any effect on the final convergence (Will the optimizer or something like this be correctly loaded?).
You should be able to see the lr curve, and as long as it is OK, the optimizer should also be OK.

>>>Why do you suggest us to try the model with bs=256 and bs=512 and they converge faster epoch-wise?
If you do not have enough GPUs using high number of grad accumulation would not help you speed-wise significantly as it just skips some of the updates. But smaller batch sizes would do more updates to the model in one epoch and it may(?) help you to converge to better numbers in the same number of epochs specially at the initial epochs. For instance bs=256 converges faster before 400 epochs than bs=2K but they both get to the same WER after that. We did most of our experiment on LS with bs=2K. I think it is safer to use bs=2K to make sure you get the same results.

>>> Does this mean that the Model with bs=512/256 and without SP will get to 8.8%/3.4% in less than 1860 epochs ?
The runs with bs=256 and bs=512 have not reached to that number of epochs yet but bs=2K without SP got to 9.1% at epoch 1800. Looks like without SP you need more than that.

>>>As described in question 2, my model with bs=256 without SP seems get worse result than your Model with bs=256 and without SP (at 460 epochs they are similar, but worse at 1000 epoch. Though the evaluation dataset is different dev clean vs test clean, I think they should be similar)
Your WER should not get worse, but better over time. If that happens, it is likely because of that bug in the SpecAug. I think that bug limited the maskings to just the first 10% of the audio. SpecAug plays a crucial role in preventing the overfitting.

>>>could you attach your config file for the 4 experiments if possible so that I could compare mine with yours?
I used the exact config file we already have with the following updates (for training with SP):
python ./examples/asr/speech_to_text$SCRIPT_POSTFIX.py  \
--config-path=./conf/conformer/ \
--config-name=conformer_ctc_bpe.yaml \
trainer.log_every_n_steps=100 \
trainer.progress_bar_refresh_rate=100 \
+trainer.benchmark=false \
model.tokenizer.type=bpe \
model.tokenizer.dir=$TOKENIZER_DIR \
model.encoder.d_model=176 \
model.encoder.n_layers=16 \
model.encoder.n_heads=4
...

for training without SP, you need to set max_epochs to 3000 epochs.

>>>BTW, for the specAug bug, do you suggest to stop the current training one and start training the model with the bug fixed one or we just continue the training with the bug ?(as I don't have many GPUs )
Suggest you just update the code and continue the training with the current checkpoint. It is better than starting from scratch.

In summary, it looks like if you want to reproduce the exact WER as ours, you need to use sentencepiece unigram tokenizer with effective batch size of 2K and train with SP for at least 650 epochs.

**Comment by BuaaAlban** [comment]
Hi， a little update, I have reached the desired WER with your help and guide!

I resume the training using the model trained  on librispeech without sp (with SpecAugment bug), which has WER= 0.0439 @282epochs on dev clean. 
And finetune it on librispeech with sp (with the bug fixed) till 339 epochs, and it reach wer=0.034 on test clean and 0.082 on test other, which is even better than the released model.

I think the key of the improvement is fixing the SpecAug bug, so I didn't train the previous model with bs 256 without sp (with the bug, it doesn't have any improvement since 500 epochs till about 2000 epochs )

**Comment by titu1994** [comment]
The spec augment kernel shouldn't even run without the correct Numba installation. But if it's the case then r1.5.1 release has the bug still on it, and the next release is in February (but fixed on main). This might be quite problematic. 

I would ask you to try to train the model from the checkpoint by forced disabling the numba spec augment by using the flag https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/modules/audio_preprocessing.py#L471 as False and finetuning the model.

**Comment by titu1994** [comment]
Or if you don't want to train force Nemo logging to debug mode with NEMO_TESTING=1 when calling the script and check for https://github.com/NVIDIA/NeMo/blob/r1.5.1/nemo/collections/asr/parts/numba/spec_augment/spec_aug_numba.py#L225 to be printed in your current run.

**Comment by BuaaAlban** [comment]
> The spec augment kernel shouldn't even run without the correct Numba installation. But if it's the case then r1.5.1 release has the bug still on it, and the next release is in February (but fixed on main). This might be quite problematic.
> 
> I would ask you to try to train the model from the checkpoint by forced disabling the numba spec augment by using the flag https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/modules/audio_preprocessing.py#L471 as False.

I didn't reinstall NeMo,  I just change the code (1.5.0)  time_starts = torch.cat(time_starts, 0) in the conda env. The code indeed use the numba kernel instead of the other one.

**Comment by titu1994** [comment]
Got it, thanks for the info. I will add it to the release notes then and it should be fixed in the next release.

**Comment by BuaaAlban** [comment]
Update result on librispeech_SP with bs 2048

epoch 407  
test clean:  WER = 0.0310 
test other: WER =0.0758

**Comment by VahidooX** [comment]
That is very interesting! I think I used the fixed code. I would check it again to make sure. Have you tried without sp?
How did you create the sp version?

**Comment by BuaaAlban** [comment]
I haven't try it without sp.  Make sure to reinstall after fixing the code or to fix the installed code directly.
My sp is just adding speed 0.9 and 1.1 using sox.
(cbn =sox.Transformer().speed(factor=s).convert(samplerate=target_sr,n_channels=1))

**Comment by VahidooX** [comment]
When you say epoch 407, do you mean 407*3=1221 epochs on the original size of LS as you used a dataset 3 times bigger?

**Comment by BuaaAlban** [comment]
Yes, but strictly speaking, the epoch 407 equals to 282 + (407-282)*3  epochs on the original size of LS as I only finetuned on a dataset 3 times bigger, and the pretrained 282 epochs are on the original size of LS
